crates/proof-of-sql/src/base/math/non_negative_i32.rs

```diff
@@ -0,0 +1,48 @@
+use alloc::fmt::Display;
+use serde::{Deserialize, Serialize};
+
+/// Type-safe non-negative integer, exists for the sole purpose
+/// of preventing negative values from being used as fixed-size
+/// binary slice widths, due to arrow's unfortunate use of i32 as width.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub struct NonNegativeI32(i32);
+
+/// Error type for `NonNegativeI32::new`.
+#[derive(Debug)]
+pub enum WidthError {
+    /// The width was negative.
+    NegativeWidth(i32),
+}
+
+impl NonNegativeI32 {
+    /// Returns an error if `x` is negative. Otherwise returns the wrapped value.
+    pub fn new(x: i32) -> Result<Self, WidthError> {
+        if x < 0 {
+            Err(WidthError::NegativeWidth(x))
+        } else {
+            Ok(Self(x))
+        }
+    }
```

Comment:
This should be named `try_new`.
We should also have `impl TryFrom<i32> for NonNegativeI32`.
Or perhaps we should remove the `try_new` altogether and only have the `TryFrom`.